### PR TITLE
Add list folding to parser

### PR DESCRIPTION
**Pull request by tohrnii**
Partly addressing #87. This PR adds support for list folding to the parser. This PR should be merged after #120 and #119.

### CODE REVIEW

File: parser/src/parser/grammar.lalrpop [hunk 1]
```diff
@@ -345,6 +348,33 @@ ListComprehension<T>: ListComprehension = {
...
+            (Members::Single(..), Iterables::Double(..)) => Err(ParseError::User {
+                error: Error::ParseError(
+                    InvalidListFolding("For a single member, a single iterable is required".to_string())
+                )
+            }),
+            (Members::Single(..), Iterables::Range(..)) => Err(ParseError::User {
+                error: Error::ParseError(
+                    InvalidListFolding("For a single member, a single iterable is required".to_string())
+                )
+            }),
```
Comment (grjte) [inline parser/src/parser/grammar.lalrpop:368]: can these be combined into a single match arm?

File: parser/src/parser/tests/list_comprehension.rs [hunk 2]
```diff
@@ -207,6 +207,190 @@ fn bc_double_list_comprehension() {
...
+                VariableType::Scalar(ListFolding(ListFoldingType::Sum(
+                    air_script_core::ListFolding::new(
+                        Box::new(Exp(
+                            Box::new(Elem(Identifier("col".to_string()))),
+                            Box::new(Const(7)),
+                        )),
+                        Members::Single(Identifier("col".to_string())),
+                        Iterables::Single(Identifier("c".to_string())),
+                    ),
+                ))),
```
Comment (grjte) [inline parser/src/parser/tests/list_comprehension.rs:242]: I guess I'm a bit confused why this isn't a `ListFolding` around a `ListComprehension`. Maybe this review should wait until the other two are merged?

**Review by grjte — COMMENTED** [review]
preliminarily I think this looks pretty good, but maybe we should wait until the other 2 are merged before reviewing

**Comment by grjte** [comment]
I think maybe we should put it in draft mode for the moment. wdyt?

File: parser/src/parser/tests/list_comprehension.rs (hunk 2 above)
Comment (tohrnii) [reply parser/src/parser/tests/list_comprehension.rs:242]: Ah, you are right, it should just be `ListFolding` around `ListComprehension`. Not sure what I was thinking there. Thanks.

File: parser/src/parser/grammar.lalrpop (hunk 1 above)
Comment (tohrnii) [reply parser/src/parser/grammar.lalrpop]: Thanks. Done.

**Review by Overcastan — APPROVED** [review]
Looks good to me, thank you @tohrnii!

**Review by grjte — APPROVED** [review]
This looks great @tohrnii, nice work!

**Pull request merged**